### DISCUSSION TITLE
Numerical oscillation?

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#23475] [Q&A General] [answered]
Dear all,

For the advection-diffusion equation,

I find two versions (compressible and incompressible versions) of advection terms in the advection-diffusion equation. 

Now I am trying to test these two cases in MOOSE. The compressible one produces stable results, however, the compressible results are unstable.

Do you have any comments or ideas about this? How to fix this? Thanks. 

J

**Comment by GiudGiud** [comment, marked as answer]
Hello

There are a few other differences here
- eq 1 has a source term R which equation 2 does not have
- eq 1 can have a spatial dependent diffusion coefficient, in eq 2 it s out of the laplacian
- eq 2 has a multiplier on the derivative, which would make the 2 equations very different

We'd need more details to conclude, notably which finite element family and order you are using and which other equations you are solving. Compressible flow is more complex for sure. 

Guillaume

**Reply by Joseph-0123** [reply]
> Hello
> 
> There are a few other differences here
> 
> * eq 1 has a source term R which equation 2 does not have
> * eq 1 can have a spatial dependent diffusion coefficient, in eq 2 it s out of the laplacian
> * eq 2 has a multiplier on the derivative, which would make the 2 equations very different
> 
> We'd need more details to conclude, notably which finite element family and order you are using and which other equations you are solving. Compressible flow is more complex for sure.
> 
> Guillaume

Dear Guillaume,

please have a look at the updated version of my problem. All the equations are the same except for the different weak forms. The main variable is ‘C'.
```
[Variables]
  [C]
    order = FIRST
    family = LAGRANGE
  []
[]
```

**Reply by GiudGiud** [reply]
you also performed an integration by parts to go from one formulation to the other. 
Did you change the boundary condition appropriately?

**Reply by Joseph-0123** [reply]
> you also performed an integration by parts to go from one formulation to the other. Did you change the boundary condition appropriately?

The compressible version has stable results, but the incompressible version with DirichletBC and NeumannBC all produce unstable results.
```
DirichletBC:
  [./left_C]
    type = DirichletBC
    variable = C
    boundary = 'left'
    value = 1
  [../]
```
or
NeumannBC
```
  [./left_C]
    type = NeumannBC
    variable = C
    boundary = 'left'
    value = 0.1 // or 0.1; or 0.01
  [../]
```

**Reply by GiudGiud** [reply]
You need to do more than that. Have a look at the Navier Stokes module finite element treatment of the pressure gradient.

I think the sign should have flipped at the very least.

Is this XY or RZ?

**Reply by Joseph-0123** [reply]
> You need to do more than that. Have a look at the Navier Stokes module finite element treatment of the pressure gradient.
> 
> I think the sign should have flipped at the very least.
> 
> Is this XY or RZ?

This is XY.

**Reply by Joseph-0123** [reply]
> You need to do more than that. Have a look at the Navier Stokes module finite element treatment of the pressure gradient.
> 
> I think the sign should have flipped at the very least.
> 
> Is this XY or RZ?

Thanks a lot, Guillaume. It works after I flip the sign.

**Reply by Joseph-0123** [reply]
> You need to do more than that. Have a look at the Navier Stokes module finite element treatment of the pressure gradient.
> 
> I think the sign should have flipped at the very least.
> 
> Is this XY or RZ?

Hello Guillaume. 

After I performed an integration by parts on the advection term, it will also produce a boundary term? How to set or deal with this boundary term? Should I just ignore it and use the default in MOOSE?

**Reply by GiudGiud** [reply]
You ll need to create a BC object for it

**Reply by Joseph-0123** [reply]
> You ll need to create a BC object for it

1. Can I just use NeumannBC for the variable C shown in the upper figures instead of creating a BC object for the advection term of variable C?

2. In addition, if you look at[ this open access app](https://github.com/ajacquey/golem/blob/master/src/bcs/GolemConvectiveTHBC.C), they use this Object BC but do not give any value of the variable in their [input file](https://github.com/ajacquey/golem/blob/master/test/tests/THM/THM_1D_transient.i)?  Does this mean that this Object BC will not work?  

3. Or does this Object BC corresponds to the extra term derived from integration by parts? Is it compulsory to use this defined  Object BC  in the input file?
